### ISSUE TITLE
[question] One-to-many performance

### ISSUE THREAD

**Issue by japel** [balderdashy/sails-mongo#371]
Consider this scenario:

```javascript
var ModelA = {
  attributes: {
    type: {
      type: "string",
      enum: [
        "typeA",
        "typeB"
      ]
    },
    children: {
      collection: "ModelA",
      via: "parent"
    },
    parent: {
      model: "ModelA"
    }
  }
};
```

So lets say I have 50k documents of type "typeA" and 5k of type "typeB".
I've seen a great performance difference between this two operations which practically end in the same result:

Operation1:
```javascript
    ModelA.find({type: "typeA"})
      .exec(function (error, found) {
        var cursor = Q.when();
        for (var i = 0, len = found.length; i < len; i++) {
          (handleOne)(found[i]);
        }

        cursor.then(function () {
          console.log("all done");
        });

        function handleOne(one) {
          cursor = cursor.then(function () {
            var d = Q.defer();
            ModelA.findOne({type: "typeB", someId: one.someId})
              .exec(function (error, typeb) {
                if (error) {
                  console.log(error);
                }
                else if (!typeb) {
                  console.log("typeb not found");
                }
                else if (typeb) {
                  typeb.children.add(one.id);
                  typeb.save(function (error, saved) {
                    if (error) {
                      console.log(error);
                    }
                    else if (saved) {
                      d.resolve();
                    }
                  });
                }
              });
            return d.promise;
          });
          return cursor;
        }
      });
```

Operation2:
```javascript
    ModelA.find({type: "typeA"})
      .exec(function (error, found) {
        var cursor = Q.when();
        for (var i = 0, len = found.length; i < len; i++) {
          (handleOne)(found[i]);
        }

        cursor.then(function () {
          console.log("all done");
        });

        function handleOne(one, i) {
          cursor = cursor.then(function () {
            var d = Q.defer();
            ModelA.findOne({type: "typeB", someId: one.someId})
              .exec(function (error, typeb) {
                if (error) {
                  console.log(error);
                }
                else if (!typeb) {
                  console.log("typeb not found");
                }
                else if (typeb) {
                  one.parent = typeb.id;
                  one.save(function (error, saved) {
                    if (error) {
                      console.log(error);
                    }
                    else if (saved) {
                      d.resolve();
                    }
                  });
                }
              });
            return d.promise;
          });
          return cursor;
        }
      });
```
Operation2 is 1000 times faster than Operation1
Could someone with deep waterline/sails-mongo undestanding tell me why this is so?

**Comment by devinivy** [comment]
To simplify your question, you're wondering why `add()`ing to a collection is not nearly as performant as directly updating the respective foreign key.  I'll put it on my todo list to look into this.  As a start, can you see if the `saved` record has associations populated in Operation1 versus Operation2?

**Comment by sailsbot** [comment]
Thanks for posting, @japel.  I'm a repo bot-- nice to meet you!

The issue queue in this repo is for verified bugs with documented features.  Unfortunately, we can't leave some other types of issues marked as "open".  This includes bug reports about undocumented features or unofficial plugins, as well as feature requests, questions, and commentary about the core framework. Please review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) for details.

If you're here looking for help and can't find it, visit [StackOverflow](http://stackoverflow.com), our [Google Group](https://groups.google.com/forum/#!forum/sailsjs) or our [chat room](https://gitter.im/balderdashy/sails?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge).  But please do not let this disrupt future conversation in this issue! I am only marking it as "closed" to keep organized.

Thanks so much for your help!

> If I am mistaken, and this issue is about a _critical bug with a documented feature_, please accept my sincerest apologies-- I am but a humble bot working to make GitHub a better place.  If you open a new issue, a member of the core team will take a look ASAP.

**Comment by devinivy** [comment]
D'oh!  This will remain open for the time being.

**Comment by japel** [comment]
@devinivy I'll take a look at it and let you know

**Comment by japel** [comment]
@devinivy I created an easy to test/debug repo with the example mentioned above:
https://github.com/japel/testParentChildRel

bootstrap.js creates a variable amount of documents for both, typeA and typeB, which you can configure here:
https://github.com/japel/testParentChildRel/blob/master/config/bootstrap.js#L20-L21
and it uses chance to create some random data (name, someMutualId)

After lifting in dev/console mode, you can either execute:
https://github.com/japel/testParentChildRel/blob/master/api/services/ParentChildRelService.js#L8
ParentChildRelService.slowOperation()
or
https://github.com/japel/testParentChildRel/blob/master/api/services/ParentChildRelService.js#L54
ParentChildRelService.fastOperation()

and it will log the time needed for it.

in my test with 5k documents of typeA and 500 documents of typeB slowOperation took 101772ms and fastOperation took 82520ms
Which at first sight might not be very a huge difference, but with larger datasets time grows at least linearly.

I hope this repo makes it for you easier to debug.

Best regards

japel

**Comment by sailsbot** [comment]
Thanks for posting, @japel.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
